### PR TITLE
Adding some, many and alt functions to Alternative

### LINKED ISSUE #1410: Add functions alt, some and many to Alternative type class
**Useful links**
* http://hackage.haskell.org/package/base-4.12.0.0/docs/Control-Applicative.html#g:2
* https://blog.jle.im/entry/free-alternative-regexp.html

### PR DESCRIPTION
**Pull request by Cotel**
Close #1410

Should we change `alt` into something else? It is supposed to be Haskell's `<|>`. I've seen some places where it is called `orElse` but it can be misleading.

I've also added a law for testing if `alt` is associative. I suppose `empty` laws are already covered with MonoidK. I took these laws from [here](https://en.wikibooks.org/wiki/Haskell/Alternative_and_MonadPlus).

I'll write some docs in another PR for completing Hacktoberfest 😋

### CODE REVIEW

File: modules/core/arrow-core/src/main/kotlin/arrow/core/extensions/listk.kt [hunk 1]
```diff
@@ -228,6 +229,9 @@ interface ListKMonadCombine : MonadCombine<ForListK> {
 
   override fun <A> Kind<ForListK, A>.combineK(y: Kind<ForListK, A>): ListK<A> =
     fix().listCombineK(y)
+
+  override fun <A> Kind<ForListK, A>.alt(b: Kind<ForListK, A>): Kind<ForListK, A> =
```
Comment (pakoito) [inline modules/core/arrow-core/src/main/kotlin/arrow/core/extensions/listk.kt]: The return can be ListK<A> due to covariance.

**Comment by pakoito** [comment]
I like `orElse`, that way we typeclass it.

File: modules/core/arrow-core-data/src/main/kotlin/arrow/typeclasses/Alternative.kt [hunk 2]
```diff
@@ -1,6 +1,16 @@
...
 /**
  * ank_macro_hierarchy(arrow.typeclasses.Alternative)
  */
-interface Alternative<F> : Applicative<F>, MonoidK<F>
+interface Alternative<F> : Applicative<F>, MonoidK<F> {
+  fun <A> Kind<F, A>.some(): Kind<F, ListK<A>> = map(this, many()) { (v, vs) -> (listOf(v) + vs).k() }
+
+  fun <A> Kind<F, A>.many(): Kind<F, ListK<A>> = some() alt just(emptyList<A>().k())
+
+  infix fun <A> Kind<F, A>.alt(b: Kind<F, A>): Kind<F, A>
```
Comment (pakoito) [inline modules/core/arrow-core-data/src/main/kotlin/arrow/typeclasses/Alternative.kt]: Please provide at least one non-infix alternative just in case :P

Comment (Cotel) [reply modules/core/arrow-core-data/src/main/kotlin/arrow/typeclasses/Alternative.kt]: Should I leave alt as the infix version and provide an orElse non-infix function? I wish we could use operators and name it <|> 😅

Comment (pakoito) [reply modules/core/arrow-core-data/src/main/kotlin/arrow/typeclasses/Alternative.kt]: ye, works for me

Comment (pakoito) [reply modules/core/arrow-core-data/src/main/kotlin/arrow/typeclasses/Alternative.kt]: One nit. Could you please swap things around so `alt` is implemented in terms of `orElse` instead?

Comment (Cotel) [reply modules/core/arrow-core-data/src/main/kotlin/arrow/typeclasses/Alternative.kt]: Sure 👍

**Pull request merged**